overlay/lib/fetch.nix

```
@@ -35,7 +35,7 @@ rec {
   fetchCrateAlternativeRegistryExpensive = { index, name, version, sha256 }: with buildPackages; stdenvNoCC.mkDerivation {
     name = "${name}-${version}.tar.gz";
 
-    inherit (fetchgitPrivate { url = git://dummy; }) GIT_SSH SSH_AUTH_SOCK;
+    inherit (builtins.fetchGit { url = git://dummy; }) GIT_SSH SSH_AUTH_SOCK;
```

This expression won't evaluate successfully since it's done at evaluation time, and `builtins.fetchGit` doesn't have `GIT_SSH` nor `SSH_AUTH_SOCK` attributes, AFAIK.

We need to either load `GIT_SSH` and `SSH_AUTH_SOCK` explicitly using `builtins.getEnv` or rewrite `fetchCrateAlternativeRegistryExpensive` to fetch the registry index using `builtins.fetchGit` properly. CC @trha

---

@macs-overdrv I think for the sake of merging this PR quickly to unblock 99% of users, we should just use `builtins.getEnv` to restore the current behavior for alternative registries.

```suggestion
    GIT_SSH = builtins.getEnv "GIT_SSH";
    SSH_AUTH_SOCK = builtins.getEnv "SSH_AUTH_SOCK";
```

We can work on switching to `builtins.fetchGit` for alternative registries at a later point, as that is a less frequently used feature.

---

By the way this cannot be used in nixpkgs.

---

@Mic92 Thanks for the update! Is it available in the latest stable Nixpkgs, at least?

---

It's in nix that we use in nixpkgs in 20.03.

---

Gotcha, thanks. Is this a temporary or permanent measure? Could you post a link to a PR or tracking issue regarding this restriction for future readers?

---

Those builtins fetchers are not allowed when using `restrict-eval`. Which is what we use when evaluating nixpkgs: https://github.com/NixOS/ofborg/#how-does-ofborg-call-nix-build

I think this is needed to not depend on network for evaluating and to avoid import from derivation.

```
$ nix eval --option restrict-eval true '(builtins.fetchGit { url = "https://github.com/NixOS/nix"; rev = "ecd4e52a580db0b928b84c049e72472c00f788db";  })'
warning: unknown setting 'experimental-features'
error: access to URI 'https://github.com/NixOS/nix' is forbidden in restricted mode
```

---

Ah, I see what you mean. I think it's fine for fetching private dependencies from alternative registries, since those will never make their way into Nixpkgs.

However, in the case of plain Git dependencies, we cannot use `fetchgit` either, since that doesn't work with fetching from private GitHub repos. We used to use `fetchgitPrivate` for this, but since that has been replaced with `builtins.fetchGit`, this is our only option left. Does there exist an alternative path?

---

I guess it should be fine than.

---

That is indeed valuable to know. Thank you! I wonder if maybe there's a heuristic we could apply which defaults to `fetchgit` and only switches to `builtins.fetchGit` if absolutely necessary? 🤔